src/protean/core/entity.py

```diff
@@ -89,6 +89,178 @@ def has_auto_field(self):
                     _, field_obj in self.declared_fields.items()])
 
 
+class QuerySet:
+    """A chainable class to gather a bunch of criteria and preferences (page size, order etc.)
+    before execution.
+
+    Internally, a QuerySet can be constructed, filtered, sliced, and generally passed around
+    without actually fetching data. No data fetch actually occurs until you do something
+    to evaluate the queryset.
+    """
+
+    def __init__(self, entity_cls_name: str, page: int = 1, per_page: int = 10,
+                 order_by: set = None, excludes_: dict = None, **filters):
+        """Initialize either with empty preferences (when invoked on an Entity)
+            or carry forward filters and preferences when chained
+        """
+
+        self._entity_cls_name = entity_cls_name
+        self._page = page or 1
+        self._per_page = per_page or 10
+
+        # `order_by` could be empty, or a string or a set.
+        #   Intialize empty set if `order_by` is None
+        #   Convert string to set if `order_by` is a String
+        #   Safe-cast set to a set if `order_by` is already a set
+        if order_by:
+            self._order_by = set([order_by]) if isinstance(order_by, str) else set(order_by)
+        else:
+            self._order_by = set()
+        self._excludes = excludes_ or {}
+        self._filters = filters
+
+    def _clone(self):
+        """
+        Return a copy of the current QuerySet.
+        """
+        clone = self.__class__(self._entity_cls_name,
+                               page=self._page, per_page=self._per_page,
+                               order_by=self._order_by, excludes_=self._excludes,
+                               **self._filters)
+        return clone
+
+    def filter(self, **filters):
+        """Merge new filter list with existing filters"""
+        clone = self._clone()
+        clone._filters.update(filters)
+
+        return clone
+
+    def exclude(self, **excludes):
+        """Merge new exclude list with existing excludes dictionary"""
+        clone = self._clone()
+        clone._excludes.update(excludes)
+
+        return clone
+
+    def page(self, page: int):
```

Comment:
The `page` and `per_page` functions case be combined into a `paginate` function

Reply:
Kept it as it is to stay in line with the most popular implementations. We can change later, if necessary.

Reply:
What do you mean by popular implementations... where have you seen this?

Reply:
I realized most of the implementations I know about come from the Rails world, and even there, they are implemented as part of gems, not as part of the core rails repo. I have moved the two methods to a combined `paginate` method now, though I am not really sure if this is correct, or needs to be done right now. Pushed as part of https://github.com/proteanhq/protean/commit/7a889a86ef17a08eb1dfedcd4525a665ec6c5844